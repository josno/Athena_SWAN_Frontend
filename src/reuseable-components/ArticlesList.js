import React from 'react'
import { Link } from 'react-router-dom'
import styled from 'styled-components'
import Card from 'react-bootstrap/card'
import Utilities from '../Utils/Utilities'
import { BsHeart } from 'react-icons/bs'

//this component allows me to create a resable articles list page
//It also gives a breif extract of the page

function slugify(content) {
  return content
    .toLowerCase()
    .replace(/ /g, '-')
    .replace(/[^\w-]+/g, '')
}

const ArticlesList = ({ articles, handleLikes, feature }) => (
  <ArticlesContainerStyles>
    {articles.map((article, key) => (
      <ArticleCardStyles key={key}>
        <Card.Img variant="top" src={article.photoUrl} />

        <Card.Body>
          <Link
            className="article-list-item"
            to={`/article/${slugify(article.name)}/${article.id}`}
          >
            <Card.Title>{article.title}</Card.Title>
            <Card.Subtitle className="mb-2 text-muted">
              {article.writtenBy} | {Utilities.convertDate(article.createdDate)}{' '}
            </Card.Subtitle>
            <Card.Text>{article.content[0].substring(0, 150)}..</Card.Text>
          </Link>
          {feature && (
            <button
              onClick={() => handleLikes(article.id, article.likes)}
              className="likes-btn"
            >
              <BsHeart className="likes-icon" size={30} />
              <span>{article.likes}</span>
            </button>
          )}
        </Card.Body>
      </ArticleCardStyles>
    ))}
  </ArticlesContainerStyles>
)
const ArticlesContainerStyles = styled.div`
  padding: 1rem;
  padding-bottom: 2rem;
`

const ArticleCardStyles = styled(Card)`
  margin: 2rem 0 2rem 0;
  border: 0px;

  .card-body {
    padding: 1.25rem 0;
  }

  .likes-btn {
    z-index: 99;
    background: transparent;
    border: none;
    color: black;
    margin: 0.5rem 0.5rem 0.5rem 0;
    padding: 0;
    span {
      margin: 0 0.5rem;
    }
    transition: 0.2s;
    :hover {
      transform: scale(1.1);
    }
  }

  @media (min-width: 720px) {
    flex-direction: row;
    img {
      width: 300px;
      max-height: 200px;
      object-fit: cover;
    }
    .card-title {
      font-size: 1.5rem;
    }
    .card-body {
      padding: 0 1.25rem;
    }
  }
`
export default ArticlesList
