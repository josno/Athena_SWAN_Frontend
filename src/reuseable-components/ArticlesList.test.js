import { shallow, mount } from 'enzyme'
import ArticlesList from './ArticlesList'
import toJson from 'enzyme-to-json'
import { BrowserRouter as Router } from 'react-router-dom'

describe('ArticlesList', () => {
  it('renders Articles list component without crashing', () => {
    const articles = [
      {
        name: 'One Love',
        id: '1tlr8HJfbi42ml7WxXpk',
        createdDate: {
          seconds: 1621947600,
          nanoseconds: '',
        },
        writtenBy: 'Bob Marley',
        likes: 30,
        content: ['lorem', 'ipsum'],
        photoUrl:
          'https://images.unsplash.com/photo-1490730141103-6cac27aaab94?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=1500&q=80',
      },
    ]
    shallow(
      <Router>
        <ArticlesList articles={articles} feature={true} />
      </Router>
    )
  })
  it('matches the snapshot', () => {
    const articles = [
      {
        name: 'One Love',
        id: '1tlr8HJfbi42ml7WxXpk',
        createdDate: {
          seconds: 1621947600,
          nanoseconds: '',
        },
        writtenBy: 'Bob Marley',
        likes: 30,
        content: ['lorem', 'ipsum'],
        photoUrl:
          'https://images.unsplash.com/photo-1490730141103-6cac27aaab94?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=1500&q=80',
      },
    ]
    const wrapper = mount(
      <Router>
        <ArticlesList articles={articles} feature={true} />
      </Router>
    )
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
