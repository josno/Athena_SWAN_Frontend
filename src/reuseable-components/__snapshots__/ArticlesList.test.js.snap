// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticlesList matches the snapshot 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <ArticlesList
      articles={
        Array [
          Object {
            "content": Array [
              "lorem",
              "ipsum",
            ],
            "createdDate": Object {
              "nanoseconds": "",
              "seconds": 1621947600,
            },
            "id": "1tlr8HJfbi42ml7WxXpk",
            "likes": 30,
            "name": "One Love",
            "photoUrl": "https://images.unsplash.com/photo-1490730141103-6cac27aaab94?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=1500&q=80",
            "writtenBy": "Bob Marley",
          },
        ]
      }
      feature={true}
    >
      <styled.div>
        <div
          className="sc-bdnxRM dimPrw"
        >
          <Styled(Card)
            key="0"
          >
            <Card
              body={false}
              className="sc-gtsrHT iJHCcq"
            >
              <div
                className="sc-gtsrHT iJHCcq card"
              >
                <CardImg
                  src="https://images.unsplash.com/photo-1490730141103-6cac27aaab94?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=1500&q=80"
                  variant="top"
                >
                  <img
                    className="card-img-top"
                    src="https://images.unsplash.com/photo-1490730141103-6cac27aaab94?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=1500&q=80"
                  />
                </CardImg>
                <CardBody>
                  <div
                    className="card-body"
                  >
                    <Link
                      className="article-list-item"
                      to="/article/one-love/1tlr8HJfbi42ml7WxXpk"
                    >
                      <LinkAnchor
                        className="article-list-item"
                        href="/article/one-love/1tlr8HJfbi42ml7WxXpk"
                        navigate={[Function]}
                      >
                        <a
                          className="article-list-item"
                          href="/article/one-love/1tlr8HJfbi42ml7WxXpk"
                          onClick={[Function]}
                        >
                          <CardTitle>
                            <ForwardRef
                              className="card-title"
                            >
                              <div
                                className="card-title h5"
                              />
                            </ForwardRef>
                          </CardTitle>
                          <CardSubtitle
                            className="mb-2 text-muted"
                          >
                            <ForwardRef
                              className="mb-2 text-muted card-subtitle"
                            >
                              <div
                                className="mb-2 text-muted card-subtitle h6"
                              >
                                Bob Marley
                                 | 
                                Tuesday, May 25, 2021
                                 
                              </div>
                            </ForwardRef>
                          </CardSubtitle>
                          <CardText>
                            <p
                              className="card-text"
                            >
                              lorem
                              ..
                            </p>
                          </CardText>
                        </a>
                      </LinkAnchor>
                    </Link>
                    <button
                      className="likes-btn"
                      onClick={[Function]}
                    >
                      <BsHeart
                        className="likes-icon"
                        size={30}
                      >
                        <IconBase
                          attr={
                            Object {
                              "fill": "currentColor",
                              "viewBox": "0 0 16 16",
                            }
                          }
                          className="likes-icon"
                          size={30}
                        >
                          <svg
                            className="likes-icon"
                            fill="currentColor"
                            height={30}
                            stroke="currentColor"
                            strokeWidth="0"
                            style={
                              Object {
                                "color": undefined,
                              }
                            }
                            viewBox="0 0 16 16"
                            width={30}
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              clipRule="evenodd"
                              d="M8 2.748l-.717-.737C5.6.281 2.514.878 1.4 3.053c-.523 1.023-.641 2.5.314 4.385.92 1.815 2.834 3.989 6.286 6.357 3.452-2.368 5.365-4.542 6.286-6.357.955-1.886.838-3.362.314-4.385C13.486.878 10.4.28 8.717 2.01L8 2.748zM8 15C-7.333 4.868 3.279-3.04 7.824 1.143c.06.055.119.112.176.171a3.12 3.12 0 01.176-.17C12.72-3.042 23.333 4.867 8 15z"
                              fillRule="evenodd"
                              key="0"
                            />
                          </svg>
                        </IconBase>
                      </BsHeart>
                      <span>
                        30
                      </span>
                    </button>
                  </div>
                </CardBody>
              </div>
            </Card>
          </Styled(Card)>
        </div>
      </styled.div>
    </ArticlesList>
  </Router>
</BrowserRouter>
`;
